src/Microsoft.Azure.WebJobs.Host/Scale/ConcurrencyManager.cs

```diff
@@ -0,0 +1,228 @@
+using Microsoft.Extensions.Options;
+
+namespace Microsoft.Azure.WebJobs.Host.Scale
+{
+    /// <summary>
+    /// Used to implement collaborative dynamic concurrency management between the host and function triggers.
+    /// Function listeners can call <see cref="GetStatus"/> within their listener polling loops to determine the
+    /// amount of new work that can be fetched. The manager internally adjusts concurrency based on various
+    /// health heuristics.
+    /// </summary>
+    public class ConcurrencyManager
+    {
+        internal const int MinConsecutiveIncreaseLimit = 5;
+        internal const int MinConsecutiveDecreaseLimit = 3;
+
+        private readonly ILogger _logger;
+        private readonly IOptions<ConcurrencyOptions> _options;
+        private readonly IConcurrencyThrottleManager _concurrencyThrottleManager;
+
+        private ConcurrencyThrottleAggregateStatus _throttleStatus;
+
+        // for mock testing only
+        internal ConcurrencyManager()
+        {
+        }
+
+        public ConcurrencyManager(IOptions<ConcurrencyOptions> options, ILoggerFactory loggerFactory, IConcurrencyThrottleManager concurrencyThrottleManager)
+        {
+            if (options == null)
+            {
+                throw new ArgumentNullException(nameof(options));
+            }
+            if (loggerFactory == null)
+            {
+                throw new ArgumentNullException(nameof(loggerFactory));
+            }
+            if (concurrencyThrottleManager == null)
+            {
+                throw new ArgumentNullException(nameof(concurrencyThrottleManager));
+            }
+
+            _options = options;
+            _logger = loggerFactory.CreateLogger(LogCategories.Concurrency);
+            _concurrencyThrottleManager = concurrencyThrottleManager;
+
+            EffectiveCoresCount = Utility.GetEffectiveCoresCount();
+        }
+
+        /// <summary>
+        /// Gets a value indicating whether dynamic concurrency is enabled.
+        /// </summary>
+        public bool Enabled => _options.Value.DynamicConcurrencyEnabled;
+
+        /// <summary>
+        /// Gets a value indicating whether any throttles are currently enabled.
+        /// </summary>
+        internal virtual bool ThrottleEnabled => _throttleStatus?.State == ThrottleState.Enabled;
+
+        internal int EffectiveCoresCount { get; set; }
+
+        internal ConcurrentDictionary<string, ConcurrencyStatus> ConcurrencyStatuses = new ConcurrentDictionary<string, ConcurrencyStatus>(StringComparer.OrdinalIgnoreCase);
+
+        /// <summary>
+        /// Get the concurrency status for the specified function. 
+        /// </summary>
+        /// <param name="functionId">This should be the full ID, as returned by <see cref="FunctionDescriptor.ID"/>.</param>
+        /// <returns>The updated concurrency status.</returns>
+        /// <remarks>
+        /// Update shouldn't be called concurrently for the same function ID.
```

Comment:
I'm not sure I understand what this remark is talking about. There is no Update method in this class. Was this documentation meant for the consumer of this method or for the maintainer of this code?

Reply:
It's referring to this method - it was called Update at one time apparently. I'll fix the comment.